src/coreclr/jit/assertionprop.cpp

```diff
@@ -2370,7 +2337,11 @@ void Compiler::optAssertionGen(GenTree* tree)
         case GT_ARR_LENGTH:
         case GT_MDARR_LENGTH:
         case GT_MDARR_LOWER_BOUND:
-            assertionInfo = optCreateAssertion(tree->GetIndirOrArrMetaDataAddr(), nullptr, OAK_NOT_EQUAL);
+            // These indirs (esp. GT_IND and GT_STOREIND) are the most popular sources of assertions.
+            if ((tree->gtFlags & GTF_EXCEPT) != 0)
```

Comment:
Seems it might make more sense to use `IndirMayFault()` here, since this check conflates the exceptions thrown by operands with the indir itself.

Reply:
Good point, but looks like `IndirMayFault` may not be cheap. In this case it's just a heuristic to avoid creating too many assertions (and we do creates too many)

Reply:
Although, I'll check

Reply:
Addressed